pinot-broker/src/main/java/com/linkedin/pinot/broker/routing/builder/PartitionAwareRealtimeRoutingTableBuilder.java

```diff
@@ -0,0 +1,150 @@
+ * from the external view.
+ *
+ */
+public class PartitionAwareRealtimeRoutingTableBuilder extends AbstractPartitionAwareRoutingTableBuilder {
+
+  @Override
+  public void computeRoutingTableFromExternalView(String tableName, ExternalView externalView,
+      List<InstanceConfig> instanceConfigList) {
+
+    // TODO: enforce the replica group concept when possible
+    //
+    // Kafka low level does not support for the concept of replica group. However, it assigns servers sequentially
+    // when building the kafka partition table. Let's assume that we have 4 servers, 2 replicas, 4 partitions.
+    //
+    // partition | replica1 | replica2
+    //         0 |  server0 |  server1
+    //         1 |  server2 |  server3
+    //         2 |  server0 |  server1
+    //         3 |  server2 |  server3
+    //
+    // When there is no partitioning on data, we can still reduce the number of servers to route if we pick servers
+    // only in either replica1 (server0, server2) or replica2 (server1, server3). Since the change in the partition
+    // mapping is rare, this additional optimization can benefit us by bringing the concept of replica group even if
+    // it's not explicitly applied. We can fall back to the regular behavior if the kafka partition mapping changes.
+    //
+    _numReplicas = Integer.valueOf(_tableConfig.getValidationConfig().getReplicasPerPartition());
+
+    // Update the cache for the segment ZK metadata
+    for (String segment : externalView.getPartitionSet()) {
+      SegmentId segmentId = new SegmentId(segment);
+      SegmentZKMetadata segmentZKMetadata = _segmentToZkMetadataMapping.get(segmentId);
+      if (segmentZKMetadata == null || segmentZKMetadata.getPartitionMetadata() == null ||
+          segmentZKMetadata.getPartitionMetadata().getColumnPartitionMap().size() == 0) {
+        segmentZKMetadata = ZKMetadataProvider.getRealtimeSegmentZKMetadata(_propertyStore, tableName, segment);
+        if (segmentZKMetadata != null) {
+          _segmentToZkMetadataMapping.put(segmentId, segmentZKMetadata);
+        }
+      }
+    }
+
+    // Gather all segments and group them by Kafka partition, sorted by sequence number
+    Map<String, SortedSet<SegmentName>> sortedSegmentsByKafkaPartition =
+        KafkaLowLevelRoutingTableBuilderUtil.getSortedSegmentsByKafkaPartition(externalView);
+
+    // Ensure that for each Kafka partition, we have at most one Helix partition (Pinot segment) in consuming state
+    Map<String, SegmentName> allowedSegmentInConsumingStateByKafkaPartition =
+        KafkaLowLevelRoutingTableBuilderUtil.getAllowedConsumingStateSegments(externalView,
+            sortedSegmentsByKafkaPartition);
+
+    RoutingTableInstancePruner pruner = new RoutingTableInstancePruner(instanceConfigList);
+
+    // Compute segment id to replica id to server instance
+    Map<SegmentId, Map<Integer, ServerInstance>> segmentId2ServersMapping = new HashMap<>();
```

Comment:
Unrelated to this PR, but seems like SegmentId is just a wrapper around string. We should evaluate if it can be eliminated completely, to avoid unnecessary object creation.